src/common/DOMWorld.ts

```diff
@@ -464,6 +476,86 @@ export class DOMWorld {
     return queryHandler.waitFor(this, updatedSelector, options);
   }
 
+  // If multiple waitFor are set up asynchronously, we need to wait for the
+  // first one to set up the binding in the page before running the others.
+  private _settingUpBinding: Promise<void> | null = null;
+  /**
+   * @internal
+   */
+  async addBindingToContext(name: string) {
+    // Previous operation added the binding so we are done.
+    if (this._ctxBindings.has(name)) return;
+    // Wait for other operation to finish
+    if (this._settingUpBinding) {
+      await this._settingUpBinding;
+      return this.addBindingToContext(name);
```

Comment:
Would it set up the binding for the same world agan?

Reply:
No it shouldn't. The first check `this._ctxBindings.has(name)` should be true if the binding was added by the current operation, so it will return immediately.
The idea behind the recursive call is that if multiple tasks are blocked on `await this._settingUpBinding;`, then one of them will get to do the recursive call and assign a new promise to `_settingUpBinding`. When the other tasks gets to do the recursive call they are therefore blocked here again.